feat(data): Make Core Data publish events to <TopicPrefix>/<DeviceProfile>/<DeviceName>

judehung opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->


## Issue Number:
fix #2934 
fix #2995 

## What is the new behavior?
Per https://github.com/edgexfoundry/edgex-docs/blob/master/docs_src/design/adr/013-Device-Service-Events-Message-Bus.md, Core data shall publish Events to <TopicPrefix>/<DeviceProfile>/<DeviceName>.

This PR made following changes:

1. Add new property PublishTopicPrefix into MessageQueue configuration.

2. Update Core Data configuration.toml to add PublishTopicPrefix set it to:
PublishTopicPrefix = 'edgex/events' # /<device-profile-name>/<device-name> will be added to this Publish Topic prefix

3. Change V2 Add Event endpoint to be /event/{DeviceProfileName}/{DeviceName}

4. When publishing V2 Event DTO use topic PublishTopicPrefix+ /{DeviceProfileName}/{DeviceName}

5. Publish []AddEventRequestDTO instead of single EventDTO to message queue

## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x] No

## New Imports
<!-- Are there any new imports or modules? If so, what are they used for and why? -->

- [ ] Yes
- [x] No

lenny-intel commented on internal/core/data/v2/application/event.go:
```diff
@@ -19,25 +19,42 @@ import (
-	lc := container.LoggingClientFrom(dic.Get)
-
-	err = checkDevice(e.DeviceName, ctx, dic)
+func AddEvent(e models.Event, profileName string, deviceName string, ctx context.Context, dic *di.Container) (id string, err errors.EdgeX) {
```
Why do we return id, since it is now always in the Event already?

judehung replied:
Right, no need to return the event id from application layer.  Removed the id.

judehung replied:
You're right. We don't have to return the id from the application layer.

lenny-intel commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -67,8 +70,9 @@ func (ec *EventController) AddEvent(w http.ResponseWriter, r *http.Request) {
 	// map Event models to AddEventResponse DTOs
 	var addResponses []interface{}
+	var successAddEventReqDTOs []requestDTO.AddEventRequest
 	for i, e := range events {
```
I just realized constraint with this approach. All events being Added must be for the same profileName and deviceName. This constraint needs to be added to Swagger documentation for AddEvent.

judehung replied:
Swagger doc updated.

lenny-intel commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -67,8 +70,9 @@ func (ec *EventController) AddEvent(w http.ResponseWriter, r *http.Request) {
+	var successAddEventReqDTOs []requestDTO.AddEventRequest
 	for i, e := range events {
-		newId, err := application.AddEvent(e, ctx, ec.dic)
+		newId, err := application.AddEvent(e, profileName, deviceName, ctx, ec.dic)
```
Add call to application.ValidateEvent prior to call to AddEvent

judehung replied:
Added.

lenny-intel commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -26,15 +26,13 @@ import (
 // NewEventController creates and initializes an EventController
 func NewEventController(dic *di.Container) *EventController {
 	return &EventController{
-		reader: io.NewEventRequestReader(),
```
Why removed. Makes sense to only create once and use it. Looks like previous code didn't use it.

judehung replied:
Yeah, I thought previous code doesn't use this reader so simply removed it; however, you're right that we only need to create the reader once and then reuses it.  Changed the code to use reader from EventController.

judehung replied:
I thought it was redundant as previous code didn't use it; however, I agree with you that we should initialize the reader once and then reuse it.  Added the reader back to event controller.

lenny-intel commented on internal/core/data/v2/controller/http/event.go:
```diff
@@ -67,8 +70,9 @@ func (ec *EventController) AddEvent(w http.ResponseWriter, r *http.Request) {
+	var successAddEventReqDTOs []requestDTO.AddEventRequest
 	for i, e := range events {
-		newId, err := application.AddEvent(e, ctx, ec.dic)
+		newId, err := application.AddEvent(e, profileName, deviceName, ctx, ec.dic)
```
`newId` can be removed and just use `e.Id`

judehung replied:
updated.

lenny-intel commented on internal/core/data/v2/controller/http/event_test.go:
```diff
@@ -173,28 +173,30 @@ func TestAddEvent(t *testing.T) {
+		{"Invalid - Invalid Reading ValueType", []requests.AddEventRequest{invalidReadingInvalidValueType}, invalidReadingInvalidValueType.Event.ProfileName, invalidReadingInvalidValueType.Event.DeviceName, true, http.StatusBadRequest},
+		{"Invalid - No SimpleReading Value", []requests.AddEventRequest{noSimpleValue}, noSimpleValue.Event.ProfileName, noSimpleValue.Event.DeviceName, true, http.StatusBadRequest},
+		{"Invalid - No BinaryReading BinaryValue", []requests.AddEventRequest{noBinaryValue}, noBinaryValue.Event.ProfileName, noBinaryValue.Event.DeviceName, true, http.StatusBadRequest},
+		{"Invalid - No BinaryReading MediaType", []requests.AddEventRequest{noBinaryMediaType}, noBinaryMediaType.Event.ProfileName, noBinaryMediaType.Event.DeviceName, true, http.StatusBadRequest},
```
Do we need tests where ProfileName or DeviceName don't match those in the Event?

judehung replied:
Well, the protocol layer will combine multiple Events inside a request, so it may not ideal to add unit test for validating mismatch profile/device name at protocol layer.  Instead, I added corresponding unit test at application layer against func `ValidateEvent`, and we may expand the unit test once `ValidateEvent` covers more validation against an event.

lenny-intel commented on internal/core/data/v2/application/event.go:
```diff
@@ -77,21 +90,21 @@ func putEventOnQueue(evt dtos.Event, ctx context.Context, dic *di.Container) {
 	}
 
-	data, err = json.Marshal(evt)
+	data, err = json.Marshal(addEvents)
```
@cloudxxx8,  I started look at the App Services side of this change to []dto.AddEventRequest and it isn't pretty. We really need a single DTO. I think this needs more discussion. Wort case for this PR is to just add loop to publish each AddEventRequest  separately.

cloudxxx8 replied:
@lenny-intel for normal event (JSON), it's fine with me to publish both single object or array.  However, it will be a problem for CBOR, because it is originally an array and we don't want to encode again here.
We have to resolve the CBOR discussion first.

lenny-intel replied:
Yes, will bring up in this weeks DS/Core WG meetings.

lenny-intel replied:
In today's DS WG we decided (tentatively?) to simplify AddEvent to be a single AddEventRequest for both JSON and CBOR. @jpwhitemn do we need to discuss this again in Core WG, or move forward with single event now?

jpwhitemn replied:
@cloudxxx8 - if you agree with reverting to a single Event being handled a single Event request then I think we should proceed without further discussion.   If Cloud believes there are additional consequences, then we should talk about it again in Core.

cloudxxx8 replied:
I'm fine with this decision.
However, the change would be large, so I suggest we can merge this PR first.
Jude will open another issue and PR to make the change.

judehung replied:
@lenny-intel @cloudxxx8 To use a single AddEventRequest DTO for both JSON and CBOR, go-mod-core-contracts needs to be updated, and I've submitted a PR, https://github.com/edgexfoundry/go-mod-core-contracts/pull/430 ,for such change.

Once https://github.com/edgexfoundry/go-mod-core-contracts/pull/430 is merged, I could either update this PR or create a new one for using single AddEventRequest DTO.

judehung replied:
@lenny-intel @cloudxxx8 I've pushed one more commit https://github.com/edgexfoundry/edgex-go/pull/3002/commits/f2181a9cfd8cd381fbc8312e3f7096b2cceae8a5 for this PR to to use single AddEventRequest.  Swagger file is also updated to reflect the change.

judehung commented on internal/core/data/v2/application/event.go:
```diff
@@ -19,25 +19,42 @@ import (
 )
 
+func validateEvent(e models.Event, profileName string, deviceName string, ctx context.Context, dic *di.Container) errors.EdgeX {
+	if profileName == "" {
+		return errors.NewCommonEdgeX(errors.KindContractInvalid, "empty profile name not allowed", nil)
+	}
+	if deviceName == "" {
+		return errors.NewCommonEdgeX(errors.KindContractInvalid, "empty device name not allowed", nil)
+	}
```
Removed.

judehung replied:
Removed.

judehung commented on internal/core/data/config/config.go:
```diff
@@ -50,8 +50,12 @@ type MessageQueueInfo struct {
-	// Indicates the topic the data is published/subscribed
+	// Indicates the topic the data is published/subscribed, this configuration shall be removed once v1 API is
+	// deprecated.
 	Topic string
```
Added

judehung replied:
Added

judehung commented on internal/core/data/v2/application/event.go:
```diff
@@ -19,25 +19,42 @@ import (
 	"github.com/google/uuid"
 )
 
+func validateEvent(e models.Event, profileName string, deviceName string, ctx context.Context, dic *di.Container) errors.EdgeX {
```
Added

judehung commented on internal/core/data/v2/application/event.go:
```diff
@@ -19,25 +19,42 @@ import (
+	lc := container.LoggingClientFrom(dic.Get)
+
 	// Add the event and readings to the database
 	if configuration.Writable.PersistData {
```
Moved and changed `validateEvent` to `ValidateEvent`

judehung commented:
@lenny-intel All of your comments have been addressed by the second commit https://github.com/edgexfoundry/edgex-go/pull/3002/commits/66ea3a14e51e3e04d027a842f9533096d1f32a6e  Please kindly take a look.  Thanks!

lenny-intel reviewed (commented):
Code looks good, we now just need to resolve the issue of multiple AddEventRequest vs single AddEventRequest published to the MessageBus (JSON & CBOR).

judehung commented:
rebased and squashed 2 commits into 1 commit.

lenny-intel commented on internal/core/data/v2/application/event.go:
```diff
@@ -54,15 +66,11 @@ func AddEvent(e models.Event, ctx context.Context, dic *di.Container) (id string
 
-// Put event DTO on the message queue to be processed by the rules engine
-func putEventOnQueue(evt dtos.Event, ctx context.Context, dic *di.Container) {
+// PublishEvent publish incoming array of AddEventRequest through MessageClient
```
Change to `PublishEvent publishes incoming AddEventRequest through MessageClient`

judehung replied:
Updated.

lenny-intel commented on internal/core/data/v2/application/event.go:
```diff
@@ -77,21 +85,21 @@ func putEventOnQueue(evt dtos.Event, ctx context.Context, dic *di.Container) {
 		lc.Debug(fmt.Sprintf(
-			"Event Published on message queue. Topic: %s, Correlation-id: %s ",
-			configuration.MessageQueue.Topic, correlationId))
+			"Event Published on message queue. Topic: %s, Correlation-id: %s ", publishTopic, correlationId))
```
Change to `V2 API Event Published on message queue...`

judehung replied:
Updated.

lenny-intel commented on openapi/v2/core-data.yaml:
```diff
@@ -459,87 +459,59 @@ components:
-        '207':
-          description: "Indicates a multi-part response supportive of accepting multiple requests at once. The 'statusCode' property of each response in the returned array will indicate success or failure."
+        '201':
+          description: "Indicates the event has been successfully created."
```
Should this be `event has been successfully added`?  since it my not be created it persistence is off.

judehung replied:
Updated.

judehung commented:
rebased and updated code per review comments from Lenny.
@lenny-intel please kindly take a look for PR changes on https://github.com/edgexfoundry/edgex-go/pull/3002/commits/bce2c04ba03c110ab37423b79f79767cefdb6fa5

lenny-intel reviewed (approved):
LGTM
